### ISSUE TITLE
Notarization/code signing issue with Autoupdate.app

### ISSUE THREAD

**Issue by MrNoodle** [sparkle-project/Sparkle#1641]
I usually try to keep my submodules free of any codesigning-specifics, preferring to apply them by using "code sign on copy" when copying the framework into my bundle.

Unfortunately, Xcode doesn't know how to code sign auxiliary binaries in frameworks, at least with "code sign on copy". To work around it, I can add a script phase to do codesign with --deep. Except, that that fails during notarization. Even when doing --deep, it seems that codesign won't see Autoupdate.app as a proper executable as it's in the wrong location. This goes for signing and for verification, which is why this type of issue doesn't usually get found until notarization.

I believe the proper fix is to put the Autoupdate.app binary in the "proper" place. Please see my blog post on the topic here: https://www.noodlesoft.com/blog/2019/04/04/frameworks-and-auxiliary-binaries/

The fix would involve the following:
* Changing the copy files phase to copy to the Executables section, not Resources
* Add an extra Run script phase to make a symbolic link of Autoupdate.app to the top level of the framework.
* Change SUBasicUpdateDriver to use -pathForAuxiliaryExecutable instead of -pathForResource

Not sure if there are any side effects that this would cause thus me submitting this.

To be clear, this fix would still require you to add a manual codesigning script unless you choose to apply your own codesigning identity to the Sparkle project. This will still benefit both cases as now doing a deep codesign verify will actually catch problems with codesigning Autoupdate.app (and fileop contained within).

**Comment by lostminds** [comment]
If you archive and upload your app for app notarization via XCode I believe all binaries are resigned already using your developer signing identity, including the binaries in the Sparkle framework. At least for me this doesn't require any additional manual codesigning scripts. I just have the framework included with the "code sign on copy" which seems to work for running locally while developing, and then the archive and notarization upload seems to resign everything correctly for distribution. So if your goal is to get your project signed correctly for notarization you may not need this change.

**Comment by MrNoodle** [comment]
Problem is that archiving has never worked on my project, despite having Apple engineers look at it in labs at WWDC.  Executables do have a certain place in bundles and I think it's more conducive to a wider variety of workflows if they are put there.

**Comment by catlan** [comment]
For example I need to generated an Installer PKG, which doesn't fit in the Xcode archive workflow. If this would fix the notarization issue, I would appreciate it being integrated.

**Comment by zorgiepoo** [comment]
I have a run script that does this (I'm using 2.x so paths & structure are a little different and Autoupdate isn't an app bundle):

```
if [ -n "$CODE_SIGN_IDENTITY" ]
then
codesign --verbose --force -o runtime --sign "$CODE_SIGN_IDENTITY" "$TARGET_BUILD_DIR/$TARGETNAME.app/Contents/Frameworks/Sparkle.framework/Versions/A/Resources/AutoUpdate"
codesign --verbose --force --deep -o runtime --sign "$CODE_SIGN_IDENTITY" "$TARGET_BUILD_DIR/$TARGETNAME.app/Contents/Frameworks/Sparkle.framework/Versions/A/Resources/Updater.app"
fi
```

I'm sure others that do notarization do something like this if you search around (eg https://furbo.org/2019/08/16/catalina-app-notarization-and-sparkle/)

And whilst I posted #546 a long time ago, I'd be weary of risk of changing locations, and making a symlink in a run script to fix up framework structure doesn't sound much more right to me. On a tangent, other things have changed recently as well (like there's no reason to not sign executables now; you should use an ad-hoc signature at least -- and I was having trouble with workflows before that involved unsigned code I believe)

Using --deep for signing is somewhat of a hammer, which the [signing guide](https://developer.apple.com/library/archive/technotes/tn2206/_index.html) explains. I use it above in Updater.app because I'm pretty sure it has no nested containing code, but otherwise you usually ought to sign things inside-to-outside order (sometimes with different entitlements for different code); similar to what Xcode does. 

I would lastly really recommend getting the Xcode Archive workflow working for you.

**Comment by MrNoodle** [comment]
Can you give me concrete reasons why changing the location would be a bad thing? I'd say my suggestion is "more right". Autoupdate.app is an auxiliary executable and does not belong in Resources. The fact that codesign will not find it is telling. Sure, -deep is not suggested for signing, but it is recommended for verify and it won't work with the current location since it doesn't expect to find executables there.

I've seen zero reasons for not doing this and only unhelpful advice funneling towards the specific workflow without regard for exceptions that may be valid. As mentioned, Apple engineers have looked at my project and their recommendation was to build for profiling and package from there.

**Comment by zorgiepoo** [comment]
I agree the change the location change be a good thing, if it can be done without major issue.

Can you confirm if [this issue](https://github.com/sparkle-project/Sparkle/issues/546#issuecomment-116876666) no longer repros or is applicable with the test application using a custom development certificate, after you make the location change? Without requiring a developer to add a run script at least for development sake.

I don't believe adding a run script to add a symlink or using pathForAuxiliaryExecutable: API is necessary by the way (you can find it from the bundle executable's symlink-resolved URL). I was at first uncertain if adding an additional symlink to the top-level of the bundle was officially supported, but it seems like the one allowed exception from the guide though it's likely framework bundles rarely do this. It's unclear to me if using the auxiliary API for non-executables is officially supported. Let us know if a symlink adds other integration benefits I may be missing here.

On one note, codesign --deep verification works the way it was intended to not validating code in resources. There might be potential implications for development workflows regarding runtime enforcements now that helpers are in a place expected to be signed and not propagated through code-sign on copy. But at least let's make sure a common "happy" path can work; developer incorporates Sparkle in their app, they use a development certificate (not ad-hoc), development works without additional run scripts, at least not until using the Archive Organizer and sending off to notarization (although if what lostminds said is true, maybe scripts then are not even needed; worth re-checking).

**Comment by MrNoodle** [comment]
The symlink allows finding it via NSBundle's URLForAuxiliaryExecutable. It seems consistent. For app bundles, binaries are put under MacOS. For frameworks, any binary images need to be symlinked at the top level. Basically, all binaries are kept together. Autoupdate is an executable binary so should be subject to the rules of auxiliary binaries.

And I want to re-iterate that relying on archiving is not feasible or desirable for some of us. Relying on it to compensate for what is arguably sloppy procedure (putting binaries in resources) is not a good approach.

**Comment by zorgiepoo** [comment]
> And I want to re-iterate that relying on archiving is not feasible or desirable for some of us. Relying on it to compensate for what is arguably sloppy procedure (putting binaries in resources) is not a good approach.

Yes I know *you* don't use that workflow. However that "happy path" needs to be tested after making this change. As well as understanding that any past showstoppers in the prior thread we had with doing this are no longer applicable. We should ensure we don't unknowingly regress developer workflows such that developers that didn't have to add run scripts before now have to.

The auxiliary executable bit sounds reasonable.

**Comment by MrNoodle** [comment]
Just to be clear, the run script happens within the Sparkle target. The user of the framework does not need to do anything. It is totally self contained.

**Comment by zorgiepoo** [comment]
Yeah I know that part is self contained, I was wondering about additional run scripts developers may or may not be using for the helper. In the past several years ago I encountered a scenario where this change exhibited a build failure in Xcode in the codesign-on-copy step. Anyway, I just did a test run with the Test Application by just changing the helper location, and I believe this change will work well. 👍
